src/ui/public/visualize/loader/visualize_data_loader.js

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isEqual } from 'lodash';
+import { VisRequestHandlersRegistryProvider } from '../../registry/vis_request_handlers';
+import { VisResponseHandlersRegistryProvider } from '../../registry/vis_response_handlers';
+
+import {
+  isTermSizeZeroError,
+} from '../../elasticsearch_errors';
+
+import { toastNotifications } from 'ui/notify';
+
+function getHandler(from, name) {
+  if (typeof name === 'function') return name;
+  return from.find(handler => handler.name === name).handler;
+}
+
+export class VisualizeDataLoader {
+  constructor(vis, Private) {
+    this._vis = vis;
+
+    const { requestHandler, responseHandler } = this._vis.type;
+
+    const requestHandlers = Private(VisRequestHandlersRegistryProvider);
+    const responseHandlers = Private(VisResponseHandlersRegistryProvider);
+    this._requestHandler = getHandler(requestHandlers, requestHandler);
+    this._responseHandler = getHandler(responseHandlers, responseHandler);
+  }
+
+  fetch = async (props, forceFetch = false) => {
```

Comment:
I think that can use the regular function syntax `async fetch(props, forceFetch = false) {` since we don't seem to call this anywhere in a weird `this`-overwriting way.